File: apps/mobile/lib/services/prediction_engine.dart
import '../models/prediction_models.dart';

class PredictionEngine {
  /// Pure deterministic rule-based calculation function.
  /// No machine learning, 100% explainable and traceable.
  static PredictionResult calculatePredictedState({
    required String userId,
    required DateTime date,
    required List<DailyCheckin> checkins,
    required WearableData wearable,
    required int cycleDay,
  }) {
    // 1. Cycle Phase Calculation
    final cyclePhase = getCyclePhase(cycleDay);
    
    // Assign Luteal (especially days 21-28) higher baseline weight for cortisol/stress (+0.35 modifier)
    double cycleModifier = 0.0;
    if (cycleDay >= 21 && cycleDay <= 28) {
      cycleModifier = 1.0; // Late luteal: full 0.35 contribution
    } else if (cycleDay >= 15 && cycleDay < 21) {
      cycleModifier = 0.45; // Early luteal
    } else if (cycleDay >= 1 && cycleDay <= 5) {
      cycleModifier = 0.30; // Menstrual
    } else {
      cycleModifier = 0.0; // Follicular / Ovulatory
    }

    // 2. Trend Aggregation (3-Day Rolling Averages)
    final sortedCheckins = List<DailyCheckin>.from(checkins)
      ..sort((a, b) => b.date.compareTo(a.date));
    
    final recentCheckins = sortedCheckins.take(3).toList();

    double rollingStressAvg = 4.0;
    double rollingMoodAvg = 7.0;
    double rollingSleepAvg = 7.5;

    if (recentCheckins.isNotEmpty) {
      rollingStressAvg = recentCheckins.map((c) => c.stressScore).reduce((a, b) => a + b) / recentCheckins.length;
      rollingMoodAvg = recentCheckins.map((c) => c.moodScore).reduce((a, b) => a + b) / recentCheckins.length;
      rollingSleepAvg = recentCheckins.map((c) => c.sleepHours).reduce((a, b) => a + b) / recentCheckins.length;
    }

    // Calculate sleep debt: IF 3-day avg sleep < 7 hours, add +0.25 to predicted stress risk
    double sleepDebtModifier = 0.0;
    if (rollingSleepAvg < 7.0) {
      sleepDebtModifier = 1.0; // Full 0.25 contribution
    }

    // 3. Weighted Scoring Engine (Weights sum to 1.0: 0.35 + 0.40 + 0.25)
    final double normalizedStress = (rollingStressAvg / 10.0).clamp(0.0, 1.0);

    // Combined Stress Index = (Cycle Modifier * 0.35) + (3-Day Avg Stress * 0.40) + (Sleep Debt Modifier * 0.25)
    final double combinedStressIndex = (cycleModifier * 0.35) +
        (normalizedStress * 0.40) +
        (sleepDebtModifier * 0.25);

    // Wearable telemetry confidence calculation
    final double confidenceScore = (0.82 + (recentCheckins.length >= 3 ? 0.08 : 0.04) + (wearable.hrv > 50 ? 0.05 : 0.02)).clamp(0.70, 0.98);

    // 4. Output Generator: State Evaluation & Primary Driver
    String predictedState;
    String primaryDriver;
    String suggestedActionTag;

    if (combinedStressIndex > 0.65) {
      predictedState = "High Stress / High Cortisol State";
      
      final List<String> drivers = [];
      if (cycleDay >= 21) {
        drivers.add("Late Luteal Cortisol Spike (Day $cycleDay)");
      }
      if (sleepDebtModifier > 0) {
        drivers.add("3-Day Sleep Debt (${rollingSleepAvg.toStringAsFixed(1)}h avg)");
      }
      if (normalizedStress > 0.5) {
        drivers.add("Elevated Tension (${rollingStressAvg.toStringAsFixed(1)}/10)");
      }
      if (drivers.isEmpty) drivers.add("Elevated Neuroendocrine Vulnerability");

      primaryDriver = drivers.join(" + ");
      suggestedActionTag = "Cortisol Reset & Somatic Breathing";
    } else if (combinedStressIndex > 0.40 && rollingMoodAvg < 5.0) {
      predictedState = "Low Energy / Emotional Vulnerability";
      primaryDriver = "Sub-Optimal Mood Recovery (${rollingMoodAvg.toStringAsFixed(1)}/10) during ${getPhaseName(cyclePhase)} Phase";
      suggestedActionTag = "Gentle Restorative Flow";
    } else {
      predictedState = "Balanced / Peak Recovery State";
      primaryDriver = "High Parasympathetic Tone (HRV ${wearable.hrv.toInt()}ms) & Restorative Sleep (${rollingSleepAvg.toStringAsFixed(1)}h avg)";
      suggestedActionTag = "Mindful Focus & Vitality Check-in";
    }

    return PredictionResult(
      userId: userId,
      date: date,
      predictedState: predictedState,
      confidenceScore: double.parse(confidenceScore.toStringAsFixed(2)),
      primaryDriver: primaryDriver,
      suggestedActionTag: suggestedActionTag,
      cyclePhase: cyclePhase,
      cycleDay: cycleDay,
      combinedStressIndex: double.parse(combinedStressIndex.toStringAsFixed(3)),
      rollingStressAvg: double.parse(rollingStressAvg.toStringAsFixed(1)),
      rollingMoodAvg: double.parse(rollingMoodAvg.toStringAsFixed(1)),
      rollingSleepAvg: double.parse(rollingSleepAvg.toStringAsFixed(1)),
    );
  }

  static CyclePhase getCyclePhase(int cycleDay) {
    if (cycleDay >= 1 && cycleDay <= 5) return CyclePhase.menstrual;
    if (cycleDay >= 6 && cycleDay <= 13) return CyclePhase.follicular;
    if (cycleDay == 14) return CyclePhase.ovulatory;
    return CyclePhase.luteal;
  }

  static String getPhaseName(CyclePhase phase) {
    switch (phase) {
      case CyclePhase.menstrual: return "Menstrual";
      case CyclePhase.follicular: return "Follicular";
      case CyclePhase.ovulatory: return "Ovulatory";
      case CyclePhase.luteal: return "Luteal";
    }
  }
}

File: apps/admin-web/src/engine/predictionEngine.js
/**
 * Pure deterministic rule-based calculation function.
 * Computes a biological and stress forecast from check-ins, wearables, and cycle day.
 */
export function calculatePredictedState({
  userId,
  date = new Date(),
  checkins = [],
  wearable = { hrv: 65, restingHR: 62, deepSleepRatio: 0.20 },
  cycleDay = 12,
}) {
  // 1. Cycle Phase Calculation
  let cyclePhase = 'follicular';
  if (cycleDay >= 1 && cycleDay <= 5) cyclePhase = 'menstrual';
  else if (cycleDay >= 6 && cycleDay <= 13) cyclePhase = 'follicular';
  else if (cycleDay === 14) cyclePhase = 'ovulatory';
  else cyclePhase = 'luteal';

  // Assign Luteal (especially days 21-28) higher baseline weight for cortisol/stress (+0.35 modifier)
  let cycleModifier = 0.0;
  if (cycleDay >= 21 && cycleDay <= 28) {
    cycleModifier = 1.0; // Full 0.35 contribution
  } else if (cycleDay >= 15 && cycleDay < 21) {
    cycleModifier = 0.45;
  } else if (cycleDay >= 1 && cycleDay <= 5) {
    cycleModifier = 0.30;
  } else {
    cycleModifier = 0.0;
  }

  // 2. Trend Aggregation (3-day rolling averages)
  const sorted = [...checkins].sort((a, b) => new Date(b.date) - new Date(a.date));
  const recent = sorted.slice(0, 3);

  let rollingStressAvg = 4.0;
  let rollingMoodAvg = 7.0;
  let rollingSleepAvg = 7.5;

  if (recent.length > 0) {
    rollingStressAvg = recent.reduce((sum, c) => sum + c.stressScore, 0) / recent.length;
    rollingMoodAvg = recent.reduce((sum, c) => sum + c.moodScore, 0) / recent.length;
    rollingSleepAvg = recent.reduce((sum, c) => sum + c.sleepHours, 0) / recent.length;
  }

  // Sleep debt modifier: IF 3-day avg sleep < 7 hours, add +0.25 to predicted stress risk
  let sleepDebtModifier = 0.0;
  if (rollingSleepAvg < 7.0) {
    sleepDebtModifier = 1.0; // Full 0.25 contribution
  }

  // 3. Weighted Scoring Engine (Weights sum to 1.0: 0.35 + 0.40 + 0.25)
  const normalizedStress = Math.min(1.0, Math.max(0.0, rollingStressAvg / 10.0));

  // Combined Stress Index = (Cycle Modifier * 0.35) + (3-Day Avg Stress * 0.40) + (Sleep Debt Modifier * 0.25)
  const combinedStressIndex =
    (cycleModifier * 0.35) +
    (normalizedStress * 0.40) +
    (sleepDebtModifier * 0.25);

  const confidenceScore = Math.min(0.98, Math.max(0.70,
    0.82 + (recent.length >= 3 ? 0.08 : 0.04) + (wearable.hrv > 50 ? 0.05 : 0.02)
  ));

  // 4. Output Generator: State Evaluation & Primary Driver
  let predictedState;
  let primaryDriver;
  let suggestedActionTag;

  if (combinedStressIndex > 0.65) {
    predictedState = "High Stress / High Cortisol State";
    const drivers = [];
    if (cycleDay >= 21) drivers.add ? drivers.push(`Late Luteal Cortisol Spike (Day ${cycleDay})`) : null;
    if (sleepDebtModifier > 0) drivers.push(`3-Day Sleep Debt (${rollingSleepAvg.toFixed(1)}h avg)`);
    if (normalizedStress > 0.5) drivers.push(`Elevated Self-Reported Tension (${rollingStressAvg.toFixed(1)}/10)`);
    if (drivers.length === 0) drivers.push("Combined Neuroendocrine Vulnerability");

    primaryDriver = drivers.join(" + ");
    suggestedActionTag = "Cortisol Reset & Somatic Breathing";
  } else if (combinedStressIndex > 0.40 && rollingMoodAvg < 5.0) {
    predictedState = "Low Energy / Emotional Vulnerability";
    primaryDriver = `Sub-Optimal Mood Recovery (${rollingMoodAvg.toFixed(1)}/10) during ${cyclePhase.toUpperCase()} Phase`;
    suggestedActionTag = "Gentle Restorative Flow";
  } else {
    predictedState = "Balanced / Peak Recovery State";
    primaryDriver = `High Parasympathetic Tone (HRV ${Math.round(wearable.hrv)}ms) & Restorative Sleep (${rollingSleepAvg.toFixed(1)}h avg)`;
    suggestedActionTag = "Mindful Focus & Vitality Check-in";
  }

  return {
    userId,
    date,
    predictedState,
    confidenceScore: parseFloat(confidenceScore.toFixed(2)),
    primaryDriver,
    suggestedActionTag,
    cyclePhase,
    cycleDay,
    combinedStressIndex: parseFloat(combinedStressIndex.toFixed(3)),
    rollingStressAvg: parseFloat(rollingStressAvg.toFixed(1)),
    rollingMoodAvg: parseFloat(rollingMoodAvg.toFixed(1)),
    rollingSleepAvg: parseFloat(rollingSleepAvg.toFixed(1)),
  };
}
